web/modules/custom/ai_screening_project_track/src/Helper/ProjectTrackHelper.php

```diff
@@ -0,0 +1,227 @@
+   *   The key.
+   *
+   * @return bool
+   *   True if data for key exists.
+   */
+  public function hasTrackData(ProjectTrackInterface $track, string $key): bool {
+    try {
+      $data = $track->getToolData();
+
+      return array_key_exists($key, $data);
+    }
+    catch (\Exception $exception) {
+      $this->logException($exception, __METHOD__, [
+        'track' => $track,
+        'key' => $key,
+      ]);
+    }
+
+    return FALSE;
+  }
+
+  /**
+   * Add data to track.
+   *
+   * @param \Drupal\ai_screening_project_track\ProjectTrackInterface $track
+   *   The track.
+   * @param string $key
+   *   The key.
+   * @param mixed $value
+   *   The value.
+   */
+  public function setTrackData(ProjectTrackInterface $track, string $key, mixed $value): void {
+    try {
+      $track->setToolData([$key => $value] + $track->getToolData());
+    }
+    catch (\Exception $exception) {
+      $this->logException($exception, __METHOD__, [
+        'track' => $track,
+        'key' => $key,
+      ]);
+    }
+  }
+
+  /**
+   * Add a webform submission to a track.
+   */
+  public function addSubmission(
+    ProjectTrackInterface $track,
+    WebformSubmissionInterface $submission,
+  ): void {
+    try {
+      $key = $submission->getEntityTypeId() . ':' . $submission->id();
+
+      $value = [
+        'created' => $this->time->getRequestTime(),
+        'webform' => $submission->getWebform()->getElementsDecoded(),
+        'submission' => $submission->getData(),
+      ];
+
+      // @todo Store the historic data in a database table to allow for easy access and querying.
+      if (!empty($submission->getData())) {
+        $historyKey = $key . ':history';
+        $history = $this->getToolData($track, $historyKey);
+        $history[] = $value;
+        $this->setTrackData($track, $historyKey, $history);
+      }
+
+      $this->setTrackData($track, $key, $value);
+
+      $computer = $this->getTrackComputer($track, $submission);
+      $computer->compute($track, $submission);
+
+      $track->save();
+    }
+    catch (\Exception $exception) {
+      $this->logException($exception, __METHOD__, [
+        'track' => $track,
+        'submission' => $submission,
+      ]);
+    }
+  }
+
+  /**
+   * Event handler.
+   */
+  public function insert(EntityInsertEvent $event): void {
+    $this->processSubmission($event);
+  }
+
+  /**
+   * Event handler.
+   */
+  public function update(EntityUpdateEvent $event): void {
+    $this->processSubmission($event);
+  }
+
+  /**
+   * Event handler.
+   */
+  public function processSubmission(EntityInsertEvent|EntityUpdateEvent $event): void {
+    $entity = $event->getEntity();
+    if ($entity instanceof WebformSubmissionInterface) {
+      if (($track = $entity->getSourceEntity()) && $track instanceof ProjectTrackInterface) {
+        $this->addSubmission($track, $entity);
+      }
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents(): array {
```

Comment:
Should this include processSubmission since that method is described as an event handler?

Reply:
Yes, you're right. I've refactored and cleaned up the code.